clients/src/test/java/org/apache/kafka/common/security/oauthbearer/internals/secured/HttpAccessTokenRetrieverTest.java

```diff
@@ -174,33 +176,39 @@ public void testParseAccessTokenInvalidJson() {
     }
 
     @Test
-    public void testFormatAuthorizationHeader() {
-        assertAuthorizationHeader("id", "secret");
+    public void testFormatAuthorizationHeader() throws UnsupportedEncodingException {
+        assertAuthorizationHeader("id", "secret", false);
     }
 
     @Test
-    public void testFormatAuthorizationHeaderEncoding() {
+    public void testFormatAuthorizationHeaderEncoding() throws UnsupportedEncodingException {
         // See KAFKA-14496
-        assertAuthorizationHeader("SOME_RANDOM_LONG_USER_01234", "9Q|0`8i~ute-n9ksjLWb\\50\"AX@UUED5E");
+        assertAuthorizationHeader("SOME_RANDOM_LONG_USER_01234", "9Q|0`8i~ute-n9ksjLWb\\50\"AX@UUED5E", false);
+        // See KAFKA-16345
+        assertAuthorizationHeader("user!@~'", "secret-(*)!", true);
     }
 
-    private void assertAuthorizationHeader(String clientId, String clientSecret) {
+    private void assertAuthorizationHeader(String clientId, String clientSecret, boolean urlencode) throws UnsupportedEncodingException {
```

Comment:
This is a bit weird that this duplicates `HttpAccessTokenRetriever.formatAuthorizationHeader()` as this simply asserts that running the same logic produces the same output

Reply:
I think the intent here is to prevent unintentional modifications to the `HttpAccessTokenRetriever.formatAuthorizationHeader()` method that could change its resulting value. 

cc. @kirktrue

Reply:
I agree that it "is a bit weird" 😄

What about moving the `expected` value to be a parameter to `assertAuthorizationHeader()` that contains the actual, pre-computed value instead of recalculating it inside the method? 

```diff
- assertAuthorizationHeader("SOME_RANDOM_LONG_USER_01234", "9Q|0`8i~ute-n9ksjLWb\\50\"AX@UUED5E", false);
+ assertAuthorizationHeader("SOME_RANDOM_LONG_USER_01234", "9Q|0`8i~ute-n9ksjLWb\\50\"AX@UUED5E", false, "U09NRV9SQU5ET01fTE9OR19VU0VSXzAxMjM0OjlRfDBgOGl+dXRlLW45a3NqTFdiXDUwIkFYQFVVRUQ1RQ==");
- assertAuthorizationHeader("user!@~'", "secret-(*)!", true);
+ assertAuthorizationHeader("user!@~'", "secret-(*)!", true, "dXNlciUyMSU0MH46c2VjcmV0LSUyOCUyQSUyOSUyMQ==");
```

**Note**: I made a very quick attempt to generate the correct value for `expected` in the above code snippet for illustrative purposes.

Reply:
yeah, I think your suggestion is good. I've updated the code.